### PR TITLE
V0.10.0

### PR DESCRIPTION
**Pull request by aldemirenes**
Since I added support for writing raw json aside from NSQ support in PR, I thought that opening new PR with this name will be good.

### CODE REVIEW

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+// Logging
+import org.slf4j.LoggerFactory
+
+// This project
+import sinks._
+import clients._
+import StreamType._
+
+/**
+ *
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: missing scaladoc

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  nsqConfig: ElasticsearchSinkNsqConfig,
+  goodSink: Option[ISink],
+  badSink: ISink,
+  elasticsearchSender: ElasticsearchSender
+) extends Runnable {
+
+  lazy val log = LoggerFactory.getLogger(getClass())
+
+  // Use same transformer with the stdin
+  // because records are coming in same format basically
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: not sure I understand

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            case Some(gs) => gs.store(s.getSource, None, true)
+            case None => elasticsearchSender.sendToElasticsearch(List(msgStr -> s.success))
+          }
+        )
+        msg.finished()
+      }
+    }
+
+    val errorCallback = new NSQErrorCallback {
+      override def error (e: NSQException) = {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: nit: space between `error` and (

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        log.error(s"Caught throwable while consuming in the NSQ", e)
+      }
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: those curlies are unnecessary

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            case None => elasticsearchSender.sendToElasticsearch(List(msgStr -> s.success))
+          }
+        )
+        msg.finished()
+      }
+    }
+
+    val errorCallback = new NSQErrorCallback {
+      override def error (e: NSQException) = {
+        log.error(s"Caught throwable while consuming in the NSQ", e)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: if there is no string interpolation, no need for the `s`.

Let's modify this to: `s"Exception while consuming topic $topicName"`

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/PlainJsonTransformer.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+// json4s
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/PlainJsonTransformer.scala]: nit: empty line

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/PlainJsonTransformer.scala [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    }
+  }
+
+  /**
+   * Consume data from stdin/NSQ rather than Kinesis
+   *
+   * @param line Line from stdin/NSQ
+   * @return Line as an EmitterInput
+   */
+  def consumeLine(line: String): EmitterInput =
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/PlainJsonTransformer.scala]: missing override

File: examples/config.hocon.sample [hunk 8]
```diff
@@ -19,24 +19,28 @@
...
   # "kinesis" for writing bad records to Kinesis
   # "stderr" for writing bad records to stderr
+  # "NSQ" for writing bad records to NSQ    
   # "none" for ignoring bad records
   bad: {{elasticsearchBadOutputDestination}}
 }
 
 # "good" for a stream of successfully enriched events
 # "bad" for a stream of bad events
+# "PlainJson" for writing plain json
```
Comment (BenFradet) [inline examples/config.hocon.sample]: I think we need to rename this to `plain-json` to be coherent with the other options

File: examples/config.hocon.sample [hunk 9]
```diff
@@ -52,6 +56,39 @@ aws {
...
+     
+  #Host name for NSQ tools
+  host: "{{NsqHost}}"
+
+  #Port for nsqd
+  port: "{{NsqdPort}}"
+
+  #Port for nsqlookupd
+  lookup-port: {{NsqlookupdPort}}
+}
```
Comment (BenFradet) [inline examples/config.hocon.sample]: It'd be cool to incorporate this configuration to a single `source` configuration similar to what has been done for stream-enrich and `streams`

File: examples/config.hocon.sample [hunk 10]
```diff
@@ -66,11 +83,6 @@ kinesis {
     maxRecords: 10000
   }
```
Comment (BenFradet) [inline examples/config.hocon.sample]: you can remove the `in` level here too

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                      .withMetricsMaxQueueSize(kinesisConnectorConfiguration.CLOUDWATCH_MAX_QUEUE_SIZE)
+                      .withUserAgent(kinesisConnectorConfiguration.APP_NAME + ","
+                              + kinesisConnectorConfiguration.CONNECTOR_DESTINATION + ","
+                              + KinesisConnectorConfiguration.KINESIS_CONNECTOR_USER_AGENT)
+                      .withRegionName(kinesisConnectorConfiguration.REGION_NAME)
+      
+      kinesisClientLibConfiguration = esLoaderConfig.kinesis.initialPosition match {
+        case "AT_TIMESTAMP" => kinesisClientLibConfiguration.withTimestampAtInitialPositionInStream(timestamp)
+        case _ => kinesisClientLibConfiguration.withInitialPositionInStream(kinesisConnectorConfiguration.INITIAL_POSITION_IN_STREAM)
+      }
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala]: let's extract all this into its own function too

File: CHANGELOG [hunk 12]
```diff
@@ -1,4 +1,10 @@
-Version 0.9.0 (2017-07-XX)
+Version 0.10.0 (2017-xx-xx)
```
Comment (BenFradet) [inline CHANGELOG]: missing line with dashes like

```text
Version 0.9.0 (2017-07-21)
--------------------------
```

File: CHANGELOG [hunk 13]
```diff
@@ -1,4 +1,10 @@
...
+Support AT_TIMESTAMP as initial position (#63)
+Configuration decoding with case-classy (#61)
```
Comment (BenFradet) [inline CHANGELOG]: pureconfig

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchEmitter.scala [hunk 14]
```diff
@@ -17,79 +17,55 @@
...
 import Scalaz._
 
-// Tracker
-import snowplow.scalatracker.Tracker
-
 // This project
 import sinks._
 import clients._
 
+
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchEmitter.scala]: empty line

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchEmitter.scala [hunk 15]
```diff
@@ -17,79 +17,55 @@
...
  * @param goodSink the configured GoodSink
  * @param badSink the configured BadSink
- * @param elasticsearchSender ES Client to use
- * @param tracker a Tracker instance
+ * @param bufferRecordLimit record limit for buffer
+ * @param bufferByteLimit byte limit for buffer
  */
-class SnowplowElasticsearchEmitter(
-  configuration: KinesisConnectorConfiguration,
+
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchEmitter.scala]: empty line

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchEmitter.scala [hunk 16]
```diff
@@ -17,79 +17,55 @@
...
-    val records = buffer.getRecords.toList
-
+  def attempEmit(records: List[EmitterInput]): List[EmitterInput] = {
     if (records.isEmpty) {
       Nil
     } else {
-
-      val (validRecords, invalidRecords) = records.partition(_._2.isSuccess)
+  
+      val (validRecords,invalidRecords) = records.partition(_._2.isSuccess)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchEmitter.scala]: missing space on this line and unnecessary spaces on the line above

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 17]
```diff
@@ -75,57 +70,110 @@ trait ElasticsearchSinkApp {
...
-    val tracker = getTracker(conf)
-
-    val executor = conf.getString("source") match {
+      
+    val executor = conf.source match {
 
       // Read records from Kinesis
-      case "kinesis" =>
-        new ElasticsearchSinkExecutor(streamType, documentIndex, documentType, finalConfig, goodSink, badSink, elasticsearchSender, tracker).success
+      case "kinesis" => new KinesisSourceExecutor(streamType, documentIndex, documentType, convertConfig(conf), conf.kinesis.initialPosition, conf.kinesis.timestamp, goodSink, badSink, elasticsearchSender, tracker).success
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: could you please split lines that are too long like this one?

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 18]
```diff
@@ -153,83 +201,16 @@ trait ElasticsearchSinkApp {
 
         // If the stream cannot be found, the KCL's "cw-metrics-publisher" thread will prevent the
         // application from exiting naturally so we explicitly call System.exit.
-        System.exit(1)
+        // This did not applied for NSQ because NSQ consumer is non-blocking and fall here
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: This does not apply to NSQ...

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+                  .withCleanupLeasesUponShardCompletion(kcc.CLEANUP_TERMINATED_SHARDS_BEFORE_EXPIRY)
+                  .withParentShardPollIntervalMillis(kcc.PARENT_SHARD_POLL_INTERVAL)
+                  .withShardSyncIntervalMillis(kcc.SHARD_SYNC_INTERVAL)
+                  .withTaskBackoffTimeMillis(kcc.BACKOFF_INTERVAL)
+                  .withMetricsBufferTimeMillis(kcc.CLOUDWATCH_BUFFER_TIME)
+                  .withMetricsMaxQueueSize(kcc.CLOUDWATCH_MAX_QUEUE_SIZE)
+                  .withUserAgent(kcc.APP_NAME + ","
+                    + kcc.CONNECTOR_DESTINATION + ","
+                    + KinesisConnectorConfiguration.KINESIS_CONNECTOR_USER_AGENT)
+                  .withRegionName(kcc.REGION_NAME)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala]: could you indent this properly?

i.e.:

```scala
val cfg = new KinesisClientLibConfiguration(
  kcc.APP_NAME,
  kcc.KINESIS_INPUT_STREAM,
  kcc.AWS_CREDENTIALS_PROVIDER,
  kcc.WORKER_ID
).withKinesisEndpoint(kcc.KINESIS_ENDPOINT)
  .withFailoverTimeMillis(kcc.FAILOVER_TIME)
  // ...
```

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+ * @param elasticsearchSender function for sending to elasticsearch 
+ * @param tracker a Tracker instance
+ */
+class KinesisSourceExecutor(
+  streamType: StreamType,
+  documentIndex: String,
+  documentType: String,
+  config: KinesisConnectorConfiguration,
+  initialPosition: String,
+  initialTimestamp: Either[String, Date],
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala]: can't you have an `Option[Date]` directly here?

Comment (BenFradet) [reply core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala]: At this point we shouldn't need the error anymore

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+        .getOrElse(cfg.withInitialPositionInStream(kcc.INITIAL_POSITION_IN_STREAM))
+  }
+
+  /**
+    * Initialize the Amazon Kinesis Client Library configuration and worker with metrics factory
+    * 
+    * @param kinesisConnectorConfiguration Amazon Kinesis connector configuration
+    * @param metricFactory would be used to emit metrics in Amazon Kinesis Client Library
+    */
+
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala]: empty line

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+    if (!kinesisConnectorConfiguration.CALL_PROCESS_RECORDS_EVEN_FOR_EMPTY_LIST) {
+      LOG.warn("The false value of callProcessRecordsEvenForEmptyList will be ignored. It must be set to true for the bufferTimeMillisecondsLimit to work correctly.")
+    }
+
+    if (kinesisConnectorConfiguration.IDLE_TIME_BETWEEN_READS > kinesisConnectorConfiguration.BUFFER_MILLISECONDS_LIMIT) {
+      LOG.warn("idleTimeBetweenReads is greater than bufferTimeMillisecondsLimit. For best results, ensure that bufferTimeMillisecondsLimit is more than or equal to idleTimeBetweenReads ")
+    }
+
+    // If a metrics factory was specified, use it.
+    if (metricFactory != null) {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/KinesisSourceExecutor.scala]: let's do:

```scala
worker = if () {
} else {
}
```

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/model.scala [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    kinesis: KinesisConfig,
+    streams: StreamsConfig,
+    elasticsearch: ESConfig,
+    monitoring: Option[MonitoringConfig]
+  ) {
+    val streamType = enabled match {
+      case "good" => Good
+      case "bad" => Bad
+      case "plain-json" => PlainJson
+      case _ => throw new RuntimeException("\"enabled\" must be set to \"good\", \"bad\" or \"plain-json\" ")
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/model.scala]: let's do `IllegalArgumentException`

File: project/Dependencies.scala [hunk 24]
```diff
@@ -22,14 +22,15 @@ object Dependencies {
...
-
+    val nsqClient        = "1.1.0-rc1"
     // Scala
     val scopt            = "3.6.0"
     val scalaz7          = "7.2.14"
     val snowplowTracker  = "0.3.0"
     val analyticsSDK     = "0.2.0"
     val awsSigner        = "0.5.0"
     val elastic4s        = "5.4.6"
+    val pureconfig       = "0.7.2"
```
Comment (BenFradet) [inline project/Dependencies.scala]: 0.8.0 is out

**Comment by BenFradet** [comment]
also there should be a rc1 commit separated from the Prepared for release commit so that you can just delete it when you release

File: core/src/test/scala/com.snowplowanalytics.elasticsearch.loader/PlainJsonTransformerSpec.scala [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics.elasticsearch.loader
+
+// Scalaz
+import scalaz._
+import Scalaz._
```
Comment (BenFradet) [inline core/src/test/scala/com.snowplowanalytics.elasticsearch.loader/PlainJsonTransformerSpec.scala]: those imports are unused

File: examples/config.hocon.sample [hunk 26]
```diff
@@ -89,47 +124,35 @@ elasticsearch {
...
 
-# Optional section for tracking endpoints
+# Section for tracking endpoints
 monitoring {
-  snowplow {
-    collector-uri: "{{collectorUri}}"
-    collector-port: 80
-    app-id: {{sinkKinesisAppName}}
-    method: GET
-  }
```
Comment (BenFradet) [inline examples/config.hocon.sample]: could you leave the snowplow bit? This makes it clearer that we're using a snowplow collector to do the monitoring

File: examples/config.hocon.sample [hunk 27]
```diff
@@ -19,25 +19,28 @@
...
+# "nsq" for reading unencoded tab-separated events from NSQ
 source = {{elasticsearchInputType}}
 
 # Where to write good and bad records
 sink {
   # Sinks currently supported are:
   # "elasticsearch" for writing good records to Elasticsearch
   # "stdout" for writing good records to stdout
-  good: {{elasticsearchGoodOutputDestination}}
+  good: "{{elasticsearchGoodOutputDestination}}"
```
Comment (BenFradet) [inline examples/config.hocon.sample]: also could we put `=` everywhere, like

```
 good = "{{elasticsearchGoodOutputDestination}}"
```

**Comment by alexanderdean** [comment]
Thanks for the detailed review @BenFradet

File: examples/config.hocon.sample [hunk 28]
```diff
@@ -48,35 +51,66 @@ stream-type: {{kinesisInStreamType}}
...
-    stream-name: {{kinesisInStreamName}} # Kinesis stream name
+  # Port for nsqd
+  port = {{nsqdPort}}
 
-    # "LATEST": most recent data.
-    # "TRIM_HORIZON": oldest available data.
-    # Note: This only affects the first run of this application on a stream.
-    initial-position: TRIM_HORIZON
+  # Port for nsqlookupd
+  lookupPort = {{nsqlookupdPort}}
```
Comment (BenFradet) [inline examples/config.hocon.sample]: we should mention that those post are the http ports not tcp

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 29]
```diff
@@ -75,57 +70,129 @@ trait ElasticsearchSinkApp {
...
+
+    val elasticsearchEndpoint = config.elasticsearch.client.endpoint
+    val elasticsearchPort = config.elasticsearch.client.port
+    val clusterName = config.elasticsearch.cluster.name
+
+    val streamRegion = config.kinesis.region
+    val appName = config.kinesis.appName.trim
+    val initialPosition = config.kinesis.initialPosition
+    val maxRecords = config.kinesis.maxRecords
+    val streamEndpoint = config.kinesis.endpoint
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: can't we remove all those vals and directly use it in the properties?

```scala
props.setProperty(KinesisConnectorConfiguration.PROP_KINESIS_INPUT_STREAM,
  config.streams.streamNameIn)
```

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  private val topicName = config.streams.streamNameIn
+  private val channelName = config.nsq.channelName
+
+ /**
+   * Consumer will be started to wait new message.
+   */
+  override def run(): Unit = {
+    val nsqCallback = new  NSQMessageCallback {
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: 2 white spaces between new and NSQ

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+      val nsqBufferSize = config.streams.buffer.recordLimit
+  
+      override def message(msg: NSQMessage): Unit = {
+        val msgStr = new String(msg.getMessage(), UTF_8)
+        msgBuffer.synchronized {
+          val emitterInput = transformer.consumeLine(msgStr)
+          msgBuffer += emitterInput
+          msg.finished()
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/NsqSourceExecutor.scala]: please remove trailing whitespaces

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/utils.scala [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+    def fold[A, B](t: Try[A])(ft: Throwable => B, fa: A => B): B = t match {
+      case Success(a) => fa(a)
+      case Failure(t) => ft(t)
+    }
+
+    def filterOrElse[L, R](e: Either[L, R])(p: R => Boolean, l: => L): Either[L, R] = e match {
+      case Right(r) if p(r) => Right(r)
+      case Right(_)         => Left(l)
+      case o                => o
+    }
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/utils.scala]: I don't think this one is used anywhere

File: examples/config.hocon.sample [hunk 33]
```diff
@@ -48,35 +51,66 @@ stream-type: {{kinesisInStreamType}}
...
-  # "app-name" is used for a DynamoDB table to maintain stream state.
+  # "appName" is used for a DynamoDB table to maintain stream state.
   # You can set it automatically using: "SnowplowElasticsearchSink-${sink.kinesis.in.stream-name}"
-  app-name: {{kinesisAppName}}
+  appName = "{{kinesisAppName}}"
+}
+
+# Common configuration section for all stream sources
+streams {
+  streamNameIn = "{{InStreamName}}"
```
Comment (BenFradet) [inline examples/config.hocon.sample]: nit: `inStreamName`

File: examples/config.hocon.sample [hunk 34]
```diff
@@ -48,35 +51,66 @@ stream-type: {{kinesisInStreamType}}
...
+  
+  # Stream for enriched events which are rejected by Elasticsearch
+  streamNameOut = "{{OutStreamName}}"
```
Comment (BenFradet) [inline examples/config.hocon.sample]: `outStreamName`

File: CHANGELOG [hunk 35]
```diff
@@ -1,4 +1,12 @@
-Version 0.9.0 (2017-07-XX)
+Version 0.10.0 (2017-xx-xx)
+--------------------------
+Create ElasticsearchEmitter class (#65)
+Support AT_TIMESTAMP as initial position (#63)
+Configuration decoding with pureconfig (#61)
+Add support for NSQ source (#59)
+Add support for writing raw JSONs (#26)
```
Comment (BenFradet) [inline CHANGELOG]: can you add the `Fix 0.9.0 entry in CHANGELOG (#60)` line?

File: examples/config.hocon.sample [hunk 36]
```diff
@@ -48,35 +51,68 @@ stream-type: {{kinesisInStreamType}}
...
-  # "app-name" is used for a DynamoDB table to maintain stream state.
+  # "appName" is used for a DynamoDB table to maintain stream state.
   # You can set it automatically using: "SnowplowElasticsearchSink-${sink.kinesis.in.stream-name}"
-  app-name: {{kinesisAppName}}
+  appName = "{{kinesisAppName}}"
+}
+
+# Common configuration section for all stream sources
+streams {
+  inStreamName = "{{InStreamName}}"
```
Comment (BenFradet) [inline examples/config.hocon.sample]: my bad I meant:

```
inStreamName = "{{inStreamName}}"
```

without the first letter capitalized

File: build.sbt [hunk 37]
```diff
@@ -19,21 +19,23 @@ lazy val commonDependencies = Seq(
...
   // Scala (test only)
   Dependencies.Libraries.specs2
 )
 
 lazy val buildSettings = Seq(
   organization  := "com.snowplowanalytics",
   name          := "snowplow-elasticsearch-loader",
-  version       := "0.9.0",
+  version       := "0.10.0-rc2",
   description   := "Load the contents of a Kinesis stream to Elasticsearch",
```
Comment (BenFradet) [inline build.sbt]: could you change that to "Load the contents of a Kinesis stream or NSQ topic to Elasticsearch"

File: core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala [hunk 38]
```diff
@@ -75,57 +70,122 @@ trait ElasticsearchSinkApp {
       System.err.println("Empty configuration file")
       System.exit(1)
     }
-    conf
+    
+    implicit def hint[T] = ProductHint[T](ConfigFieldMapping(CamelCase, CamelCase))
+    val esLoaderConf = loadConfig[ESLoaderConfig](conf) match {
+      case Left(e) =>
+        System.err.println(s"configuration error: $e")
+        System.exit(-1)
```
Comment (BenFradet) [inline core/src/main/scala/com.snowplowanalytics.elasticsearch/loader/ElasticsearchSinkApp.scala]: it's exit(1) not -1

**Pull request closed without merging**